utils/fast-math/logs.hpp

```diff
@@ -19,50 +17,98 @@
 #include <cmath>
 #include <ports-of-call/portability.hpp>
 
+/*
+ * These functions are for use when moving to and from the gridding space
+ * for Spiner grids.
+ *
+ * The core idea here is to move onto a grid that's roughly
+ * logarithmically spaced. It doesn't actually matter if that grid is
+ * EXACTLY logarithmic. Just that it is approximately so.
+ * There are however some constraints that do matter:
+ *
+ * 1. The function we use to translate to "grid space" must be
+ * precise, meaning you get the same answer each time.
+ *
+ * 2. The function must be invertible, meaning you can go back to
+ * "linear space." Ideally it is also continuous.
+ *
+ * 3. The function and its inverse must be fast.
+ *
+ * To meet these constraints, we approximate a logarithm by using
+ * frexp and ldexp to split a real number into its mantissa + exponent
+ * in base 2. The mantissa is a real number on the interval [0.5, 1)
+ * and the exponent is an integer such that
+ *
+ * mantissa * 2^exponent = number
+ *
+ * To truly take the log, one needs to take the log of the mantissa.
+ * However, the first term in the logarithm's Taylor series is linear.
+ * Thus a linear approximation to log, for small numbers is valid.
+ * We use that linear approximation, which turns out to be:
+ *
+ * 2*(mantissa - 1)
+ *
+ * So that mantissa = 0.5 returns log_2(0.5) = -1 and mantissa (1)
+ * yields 0.
+ *
+ * Then the approximate log of the whole thing is the approximation of
+ * the log of the mantissa, plus the exponent:
+ *
+ * 2*(mantissa - 1) + exponent
+ *
+ * Which is a continuous, invertible function that approximates lg
+ * only relatively accurately. The absolute error of this
+ * approximation is about 0.1 at its largest. This translates to a
+ * relative error of at most 25%.
+ *
+ * However, we don't mind this error, as we're only interested in
```

Comment:
this is a bit better than above, but worth saying again that the error you're referring to here is not actually any error in what we're doing.